### PR TITLE
IndicesClusterStateService should clean local started when re-assigns an initializing shard with the same aid

### PR DESCRIPTION
**Pull request by bleskes**
When a node get disconnected from the cluster and rejoins during a master election, it may be that the new master already has that node in it's cluster and will try to assign it shards. If the node hosts started primaries, the new shards will be initializing and will have the same allocation id as the allocation ids of the current started size. We currently do not recognize this currently. We should clean the current `IndexShard` instances and initialize new ones.

This also hardens test assertions in the same area.

### CODE REVIEW

**Comment by bleskes** [comment]
retest this please

File: core/src/test/java/org/elasticsearch/indices/cluster/AbstractIndicesClusterStateServiceTestCase.java [hunk 1]
```diff
@@ -331,9 +337,21 @@ public IndexShardState state() {
         @Override
         public void updateRoutingEntry(ShardRouting shardRouting) throws IOException {
             failRandomly();
-            assert this.shardId().equals(shardRouting.shardId());
-            assert this.shardRouting.isSameAllocation(shardRouting);
+            assertThat(this.shardId(), equalTo(shardRouting.shardId()));
+            assertTrue("current: " + this.shardRouting + ", got: " + shardRouting, this.shardRouting.isSameAllocation(shardRouting));
+            if (this.shardRouting.active()) {
+                assertTrue("and active shard must state active, current: " + this.shardRouting + ", got: " + shardRouting,
```
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/indices/cluster/AbstractIndicesClusterStateServiceTestCase.java]: I guess this reads "an active shard must stay active"... ;-)

**Review by ywelsch — APPROVED** [review]
Left minor comments. Fix and push at will.

File: core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java [hunk 2]
```diff
@@ -370,11 +370,18 @@ private void removeShards(final ClusterState state) {
...
                     // we can just remove the shard without cleaning it locally, since we will clean it in IndicesStore
                     // once all shards are allocated
                     logger.debug("{} removing shard (not allocated)", shardId);
                     indexService.removeShard(shardId.id(), "removing shard (not allocated)");
+                } else if (newShardRouting.isSameAllocation(currentRoutingEntry) == false) {
+                    logger.debug("{} removing shard (stale allocation id, stale {}, new {})", shardId,
+                        currentRoutingEntry, newShardRouting);
+                    indexService.removeShard(shardId.id(), "removing shard (stale copy)");
+                } else if (newShardRouting.initializing() && currentRoutingEntry.active()) {
+                    logger.debug("{} removing shard (not active, current {}, new {})", shardId, currentRoutingEntry, newShardRouting);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java]: added

File: core/src/test/java/org/elasticsearch/indices/cluster/AbstractIndicesClusterStateServiceTestCase.java [hunk 3]
```diff
@@ -303,9 +308,10 @@ public Index index() {
     protected class MockIndexShard implements IndicesClusterStateService.Shard {
         private volatile ShardRouting shardRouting;
         private volatile RecoveryState recoveryState;
+        private volatile long term;
 
-        public MockIndexShard(ShardRouting shardRouting) {
-            this.shardRouting = shardRouting;
+        public MockIndexShard(ShardRouting shardRouting, long term) {
+            this.shardRouting = shardRouting; this.term = term;
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/indices/cluster/AbstractIndicesClusterStateServiceTestCase.java]: I guess it was folded by intellij into a one liner and I just added it in there. Anyway - fixed.

File: core/src/test/java/org/elasticsearch/indices/cluster/AbstractIndicesClusterStateServiceTestCase.java (hunk 1 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/indices/cluster/AbstractIndicesClusterStateServiceTestCase.java]: :)  can't get rid of these states.

**Comment by bleskes** [comment]
thx @ywelsch

**Pull request merged**